msdm/algorithms/search.py

```diff
@@ -0,0 +1,97 @@
+import collections
+import heapq
+import random
+
+from msdm.core.algorithmclasses import Plans, Result
+from msdm.core.assignment import AssignmentSet, AssignmentMap
+from msdm.core.problemclasses.mdp import DeterministicShortestPathProblem
+from msdm.core.problemclasses.mdp.policy.partialpolicy import PartialPolicy
+
+def reconstruct_path(camefrom, start, terminal_state):
+    '''
+    Reconstructs a path using a dictionary mapping states
+    to the state that preceded them.
+    '''
+    path = [terminal_state]
+    while path[-1] != start:
+        path.append(camefrom[path[-1]])
+    return path[::-1]
+
+def path_to_policy(path):
+    '''
+    Converts a path (a sequence of states from a start to a goal) into a policy.
+    '''
+    return PartialPolicy(AssignmentMap([
+        (s, AssignmentMap([(ns, 1)]))
+        for s, ns in zip(path[:-1], path[1:])
+    ]))
+
+class BreadthFirstSearch(Plans):
+    def plan_on(mdp: DeterministicShortestPathProblem):
+        start = mdp.initial_state()
+
+        queue = collections.deque([start])
+
+        visited = AssignmentSet()
+        camefrom = AssignmentMap()
+
+        while queue:
+            s = queue.popleft()
+
+            if mdp.is_terminal(s):
+                path = reconstruct_path(camefrom, start, s)
+                return Result(
+                    path=path,
+                    policy=path_to_policy(path),
+                    visited=visited,
+                )
+
+            visited.add(s)
+
+            for a in mdp.actions(s):
+                ns = mdp.next_state(s, a)
+                if ns not in visited and ns not in queue:
+                    queue.append(ns)
+                    camefrom[ns] = s
+
+class AStarSearch(Plans):
+    """
+    A* Search is an informed best-first search algorithm. It considers states in priority order
+    based on the cost to reach a state and a heuristic cost.
+
+    Here, the heuristic cost is specified by a heuristic _value_ function, so a typical
+    search heuristic for the cost should be negated.
+    """
+    def plan_on(mdp: DeterministicShortestPathProblem, *, heuristic_value=lambda s: 0, seed=None):
```

Comment:
The MDP algorithms (e.g., VI) tend to be used by initializing an algorithm with a bunch of parameters and then are used to plan_on an mdp, e.g.:
```
mdp = MDP()
vi = VectorizedValueIteration()
res = vi.plan_on(mdp)
# or you could do something like
res = VectorizedValueIteration().plan_on(mdp)
```
Open to discussing if this is the best interface but would be good to keep it consistent for now.

Reply:
:+1: